### PR TITLE
747 Add distributed training example

### LINKED ISSUE #747: Add example to show distributed training
**Is your feature request related to a problem? Please describe.**
In order to show the flexibility of MONAI APIs for the complicated environment, need to investigate and develop an example for multi-nodes multi-GPUs training.
The first version can be a PyTorch DDP based on example with `nccl` backend, may also add Horovod example in future.

### PR DESCRIPTION
**Pull request by Nic-Ma**
Fixes #747 .

### Description
This PR implemented distributed training example based on the PyTorch native DDP module.
I may also add Horovod version example later referring to: https://horovod.readthedocs.io/en/latest/pytorch.html

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh --codeformat --coverage`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Comment by Nic-Ma** [comment]
Tested locally as below:
On node 0 (master):
```
apt-sh-ai:876:876 [0] NCCL INFO Bootstrap : Using [0]eno1:10.23.137.29<0>
apt-sh-ai:876:876 [0] NCCL INFO NET/Plugin : No plugin found (libnccl-net.so), using internal implementation
apt-sh-ai:876:876 [0] NCCL INFO NET/IB : No device found.
apt-sh-ai:876:876 [0] NCCL INFO NET/Socket : Using [0]eno1:10.23.137.29<0>
apt-sh-ai:876:876 [0] NCCL INFO Using network Socket
NCCL version 2.7.5+cuda11.0
apt-sh-ai:876:912 [0] NCCL INFO Channel 00/02 :    0   1
apt-sh-ai:876:912 [0] NCCL INFO Channel 01/02 :    0   1
apt-sh-ai:876:912 [0] NCCL INFO threadThresholds 8/8/64 | 16/8/64 | 8/8/64
apt-sh-ai:876:912 [0] NCCL INFO Trees [0] 1/-1/-1->0->-1|-1->0->1/-1/-1 [1] -1/-1/-1->0->1|1->0->-1/-1/-1
apt-sh-ai:876:912 [0] NCCL INFO Setting affinity for GPU 0 to ffffff
apt-sh-ai:876:912 [0] NCCL INFO Channel 00 : 1[1000] -> 0[2000] [receive] via NET/Socket/0
apt-sh-ai:876:912 [0] NCCL INFO Channel 00 : 0[2000] -> 1[1000] [send] via NET/Socket/0
apt-sh-ai:876:912 [0] NCCL INFO Channel 01 : 1[1000] -> 0[2000] [receive] via NET/Socket/0
apt-sh-ai:876:912 [0] NCCL INFO Channel 01 : 0[2000] -> 1[1000] [send] via NET/Socket/0
apt-sh-ai:876:912 [0] NCCL INFO 2 coll channels, 2 p2p channels, 1 p2p channels per peer
apt-sh-ai:876:912 [0] NCCL INFO comm 0x7f23b0006880 rank 0 nranks 2 cudaDev 0 busId 2000 - Init COMPLETE
apt-sh-ai:876:876 [0] NCCL INFO Launch mode Parallel
----------
epoch 1/5
1/20, train_loss: 0.5628
2/20, train_loss: 0.6157
3/20, train_loss: 0.5859
4/20, train_loss: 0.4808
5/20, train_loss: 0.4167
6/20, train_loss: 0.4638
7/20, train_loss: 0.4510
8/20, train_loss: 0.4581
9/20, train_loss: 0.4240
10/20, train_loss: 0.4497
epoch 1 average loss: 0.4909
----------
epoch 2/5
1/20, train_loss: 0.4379
2/20, train_loss: 0.4383
3/20, train_loss: 0.4314
4/20, train_loss: 0.4233
5/20, train_loss: 0.3747
6/20, train_loss: 0.3485
7/20, train_loss: 0.3474
```
On node 1:
```
Nic-DL:926:926 [0] NCCL INFO Bootstrap : Using [0]eno1:10.19.183.62<0>
Nic-DL:926:926 [0] NCCL INFO NET/Plugin : No plugin found (libnccl-net.so), using internal implementation
Nic-DL:926:926 [0] NCCL INFO NET/IB : No device found.
Nic-DL:926:926 [0] NCCL INFO NET/Socket : Using [0]eno1:10.19.183.62<0>
Nic-DL:926:926 [0] NCCL INFO Using network Socket
Nic-DL:926:945 [0] NCCL INFO threadThresholds 8/8/64 | 16/8/64 | 8/8/64
Nic-DL:926:945 [0] NCCL INFO Trees [0] -1/-1/-1->1->0|0->1->-1/-1/-1 [1] 0/-1/-1->1->-1|-1->1->0/-1/-1
Nic-DL:926:945 [0] NCCL INFO Channel 00 : 0[2000] -> 1[1000] [receive] via NET/Socket/0
Nic-DL:926:945 [0] NCCL INFO Channel 00 : 1[1000] -> 0[2000] [send] via NET/Socket/0
Nic-DL:926:945 [0] NCCL INFO Channel 01 : 0[2000] -> 1[1000] [receive] via NET/Socket/0
Nic-DL:926:945 [0] NCCL INFO Channel 01 : 1[1000] -> 0[2000] [send] via NET/Socket/0
Nic-DL:926:945 [0] NCCL INFO 2 coll channels, 2 p2p channels, 1 p2p channels per peer
Nic-DL:926:945 [0] NCCL INFO comm 0x7f7bac006880 rank 1 nranks 2 cudaDev 0 busId 1000 - Init COMPLETE
----------
epoch 1/5
1/20, train_loss: 0.5998
2/20, train_loss: 0.5835
3/20, train_loss: 0.4805
4/20, train_loss: 0.5042
5/20, train_loss: 0.4743
6/20, train_loss: 0.4480
7/20, train_loss: 0.4027
8/20, train_loss: 0.4351
9/20, train_loss: 0.4500
10/20, train_loss: 0.4694
epoch 1 average loss: 0.4847
----------
epoch 2/5
1/20, train_loss: 0.4658
2/20, train_loss: 0.4132
3/20, train_loss: 0.3681
4/20, train_loss: 0.5152
5/20, train_loss: 0.4558
6/20, train_loss: 0.3947
```
Every process only handles half of the iterations in an epoch.

Thanks.

File: examples/distributed_training/unet_training_ddp.py [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+    RandCropByPosNegLabeld,
+    RandRotate90d,
+    ToTensord,
+)
+from monai.data import create_test_image_3d, list_data_collate
+
+
+def train(gpu, args):
+    def _log_print(data, gpu=gpu):
+        print(data) if gpu == 0 else None
```
[the remark this answers is not in the document]
Comment (ericspod) [reply examples/distributed_training/unet_training_ddp.py:69]: For interest's sake an alternative is to redirect stdout/stderr for every process that isn't for GPU0:

```
import os
import sys
if gpu != 0:
    f = open(os.devnull, 'w')
    sys.stdout = sys.stderr = f
```

File: examples/distributed_training/unet_training_ddp.py [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+            im, seg = create_test_image_3d(128, 128, 128, num_seg_classes=1, channel_dim=-1)
+
+            n = nib.Nifti1Image(im, np.eye(4))
+            nib.save(n, os.path.join(args.dir, f"img{i:d}.nii.gz"))
+
+            n = nib.Nifti1Image(seg, np.eye(4))
+            nib.save(n, os.path.join(args.dir, f"seg{i:d}.nii.gz"))
+
+    args.world_size = args.gpus * args.nodes
+    os.environ["MASTER_ADDR"] = "10.23.137.29"
```
[the remark this answers is not in the document]
Comment (Nic-Ma) [reply examples/distributed_training/unet_training_ddp.py:179]: The address is for the master node, yes you are right, args should be better, updated in the latest commit.
Thanks.

File: examples/distributed_training/unet_training_ddp.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Nic-Ma) [reply examples/distributed_training/unet_training_ddp.py:69]: Thanks for your suggestion, updated to redict stdout/stderr in the latest commit.

File: examples/distributed_training/unet_training_ddp.py [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+    _log_print(f"train completed, epoch losses: {epoch_loss_values}")
+    if rank == 0:
+        torch.save(model.state_dict(), "final_model.pth")
+    dist.destroy_process_group()
+
+
+def main():
+    monai.config.print_config()
+    logging.basicConfig(stream=sys.stdout, level=logging.INFO)
+    set_determinism(seed=0)
```
[the remark this answers is not in the document]
Comment (Nic-Ma) [reply examples/distributed_training/unet_training_ddp.py:156]: Yes, your concern makes sense. Actually, I didn't mean to set deterministic training for this example.
I just wanted to make generate a same random dataset for every node.
Changed to use `np.random.seed` directly in the latest commit.
Thanks.

**Comment by Nic-Ma** [comment]
Hi @ericspod ,

Thanks very much for your review!
I updated the PR according to your comments and suggestions.
Could you please help review it again?

Thanks.
